src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,515 @@
+        (*domainConfigs).get_child_optional(std::to_string(cbid));
+    if (!faultConfig) {
+        faultConfig = (*domainConfigs).get_child_optional(key);
+    }
+    if (!faultConfig) {
+        faultConfig = (*domainConfigs).get_child_optional("*");
+    }
+    return faultConfig;
+}
+
+
+void CUPTIAPI
+faultInjectionCallbackHandler(
+    void *,
+    CUpti_CallbackDomain domain,
+    CUpti_CallbackId cbid,
+    void *cbdata
+) {
+    const std::string faultInjectorKernelPrefix = std::string("faultInjectorKernel");
+
+    CUpti_CallbackData *cbInfo = static_cast<CUpti_CallbackData*>(cbdata);
+
+    // TODO maybe allow it in the config but right now CUPTI_API_EXIT is generally preferrable
+    // for interception. However, it means that the execution has happened.
+    //
+    if (cbInfo->callbackSite != CUPTI_API_EXIT) {
+        return;
+    }
+
+    // Check last error
+    CUPTI_CALL(cuptiGetLastError());
+    boost::optional<boost::property_tree::ptree&> matchedFaultConfig = boost::none;
+    PTHREAD_CALL(pthread_rwlock_rdlock(&globalControl.configLock));
+
+    // check if we are processing the result of our own launch.
+    // symbolName is only valid for launches
+    //
+    // https://gitlab.com/nvidia/headers/cuda-individual/cupti/-/blob/main/cupti_driver_cbid.h
+    // https://gitlab.com/nvidia/headers/cuda-individual/cupti/-/blob/main/cupti_runtime_cbid.h
+    //
+    if (domain == CUPTI_CB_DOMAIN_DRIVER_API && globalControl.driverFaultConfigs) {
+        switch (cbid) {
+
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunch:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchGrid:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchGridAsync:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchKernel:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchKernel_ptsz:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchCooperativeKernel:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchCooperativeKernel_ptsz:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchCooperativeKernelMultiDevice:
+        case CUPTI_DRIVER_TRACE_CBID_cuGraphLaunch:
+        case CUPTI_DRIVER_TRACE_CBID_cuGraphLaunch_ptsz:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchHostFunc:
+        case CUPTI_DRIVER_TRACE_CBID_cuLaunchHostFunc_ptsz:
+        // 11.7
+        // case CUPTI_DRIVER_TRACE_CBID_cuLaunchKernelEx:
+        // case CUPTI_DRIVER_TRACE_CBID_cuLaunchKernelEx_ptsz:
+            if (std::string(cbInfo->symbolName).compare(0, faultInjectorKernelPrefix.size(), faultInjectorKernelPrefix) == 0) {
+                spdlog::debug("rejecting fake launch functionName={} symbol={}",
+                    cbInfo->functionName, cbInfo->symbolName);
+                break;
+            }
+            // intentional fallthrough
+        default:
+            matchedFaultConfig = lookupConfig(globalControl.driverFaultConfigs, cbInfo->functionName, cbid);
+        }
+    }
+    if (domain == CUPTI_CB_DOMAIN_RUNTIME_API && globalControl.runtimeFaultConfigs) {
```

Comment:
Should be `else if`.

Reply:
I find it more readable the way it's symmetrically written, but either way is fine with me in this case.